---
title: "Genes1"
author: "Enrico"
date: "24 April 2018"
output:
  html_document: default
---

#0a: Define paths.

```{r Define paths, eval=FALSE, engine='bash'}

ssh -X ebazzicalupo@genomics-a.ebd.csic.es

ssh -X ebazzicalupo@genomics-b.ebd.csic.es

# path to gff data in my laptop
GFF_Path=~/home/ebazzicalupo/Data

# path to gff data in server
GFF_Path=~/Data

```

#1a: Search in Annotations

```{r Search in Annotations, eval=FALSE, engine='bash'}

cd $GFF_Path

# list elements in file containing list of genes
GENES=$(cat GeneList.txt)

mkdir LynxResults

# define the end of a line as a separator for elements so that each WHOLE line in GeneList.txt will be a search entry
IFS=$'\n' 

#for each gene create a file containing results
for i in ${GENES[@]}
  do
  grep -i -E "${i}" LYPA23C.all.fix.gff3 > ${i}
  mv ${i} LynxResults
done


```

#1b: Refine Lynx Results - Unique Parent Gene IDs

```{r Refine Lynx Results, eval=FALSE, engine='bash'}

cd $GFF_Path/LynxResults

LYNXREFINE=$(ls *)

mkdir ParentIDs

for i in ${LYNXREFINE[@]}
  do cut -d ';' -f2 "${i}" | sort -u > "${i}".Parent
  mv "${i}".Parent ParentIDs
done


```

#2a: Search empty results in Cat annotations

```{r Search in Annotations, eval=FALSE, engine='bash'}

cd $GFF_Path

cd LynxResults

cd $GFF_Path

mkdir CatResults

# Grab empty results from Lynx search - ALTERNATIVE: without .txt file
CATGENES=$(wc -l LynxResults/* | grep ' 0 ' | cut -d '0' -f2 | sed 's/ //' | sed 's/LynxResults\///')

# define the end of a line as a separator for elements so that each WHOLE line in GeneList.txt will be a search entry
IFS=$'\n'

for i in $CATGENES; do grep -i -E "${i}" Felis_catus.Felis_catus_8.0.92.gff3 > cat_"${i}".gff3; mv cat_"${i}".gff3 CatResults; done

```

#2b: (EXAMPLE FOR 1 FILE): Generate Cat gene fasta file for BLAST - Bedtools installation required

```{r Generate Cat gene fasta file for BLAST, eval=FALSE, engine='bash'}

cd $GFF_Path

# create gene file for trial
grep "agouti" Felis_catus.Felis_catus_8.0.92.gff3 > cat_agouti.gff3

mv cat_agouti.gff3 Cat_Results

# generate fasta file (-fo) from input fasta file (-fi) using coordinates specified in other file (-bed)
bedtools getfasta -fo try.fa -fi Felis_catus.Felis_catus_8.0.dna_sm.toplevel.fa -bed cat_agouti.gff3


```

#2c: Generate Cat gene fasta files for BLAST - Bedtools installation required

```{r Generate Cat gene fasta files for BLAST, eval=FALSE, engine='bash'}

cd $GFF_Path/CatResults

# remove empty results in server

wc -l * | grep ' 0 ' | cut -d ' ' -f5 > Removelist.txt

REMlist=$(cat Removelist.txt)

for i in ${REMlist[@]}
  do
  rm ${i}
done

# Remove non-gene (e.g. mRNA) strings from .gff3 results

cd $GFF_Path/CatResults

CATREFINE=$(ls cat_*.gff3)

for i in ${CATREFINE[@]}
  do
  awk -v "key=gene" '$3 == key {print($0)}' ${i} > ${i/.gff3}.gene.gff3
done

# list all cat gene coordinates files
CATcoords=$(ls cat_*.gene.gff3)

# generate fasta file for each gene.gff3 coordinates file
for i in ${CATcoords[@]}
  do
  bedtools getfasta -fo "${i/.gene.gff3}".fa -fi $GFF_Path/Felis_catus.Felis_catus_8.0.dna_sm.toplevel.fa -bed ${i}
done

```

#2d: BLAST cat genes to Lynx genome - BLAST installation required (from NCBI)

```{r  BLAST cat genes to Lynx genome, eval=FALSE, engine='bash'}

cd $GFF_Path

# create lynx BLASTdatabase (directory and database inside) from FASTA lynx reference genome

mkdir BLASTdatabase

# in my laptop
BDB=~/home/ebazzicalupo/Data/BLASTdatabase
# in server
BDB=~/Data/BLASTdatabase

mv lp23_plus_mt.fa $BDB

makeblastdb -in $BDB/lp23_plus_mt.fa -parse_seqids -dbtype nucl

# BLAST procedure with new directory containing results file for each GENE (look at BLAST manual for output options)

cd $GFF_Path/CatResults

IFS=$'\n'

CBLASTGENES=$(ls cat_*.fa)

for GENE in ${CBLASTGENES[@]}
  do
  echo "$GENE starting BLAST"
  blastn -query $GFF_Path/CatResults/${GENE} -db $BDB/lp23_plus_mt.fa -outfmt "6 qseqid sseqid sstart send length qcovhsp evalue" -out ${GENE/.fa/}.BLASTresults
  echo "$GENE BLAST finished"
  
done

# Move BLASTresults to new directory

mkdir BLASTresults

mv *.BLASTresults BLASTresults

```

#2e: BLAST cat genes to Lynx discarded scaffold - BLAST installation required (from NCBI)

```{r BLAST cat genes to Lynx discarded scaffold, eval=FALSE, engine='bash'}

## Create BLAST database for discarded scaffold

cd $GFF_Path

mkdir Discarded_BLASTdatabase

# in my laptop
DSBDB=~/home/ebazzicalupo/Data/Discarded_BLASTdatabase
# in server
DSBDB=~/Data/Discarded_BLASTdatabase

mv lp23.discarded.scaffolds.fa $DSBDB

makeblastdb -in $DSBDB/lp23.discarded.scaffolds.fa -parse_seqids -dbtype nucl

cd $GFF_Path/CatResults

CDSBLASTGENES=$(ls cat_*.fa)
IFS=$'\n'

for GENE in ${CDSBLASTGENES[@]}
  do
  echo "$GENE starting BLAST"
  blastn -query $GFF_Path/CatResults/${GENE} -db $DSBDB/lp23.discarded.scaffolds.fa -outfmt "6 qseqid sseqid sstart send length qcovhsp evalue" -out ${GENE/.fa/}.DSBLASTresults
  echo "$GENE BLAST finished"
  
done

# Move BLASTresults to new directory

mkdir DSBLASTresults

mv *.DSBLASTresults DSBLASTresults


```


#3a: Search cat empty results in Human annotations

```{r Search cat empty results in Human annotations, eval=FALSE, engine='bash'}

cd $GFF_Path

mkdir HumanResults

# Grab empty results from Cat search - ALNTERNATIVE, with no .txt file

HUMANGENES=$(wc -l CatResults/*.gene.gff3 | grep ' 0 ' | cut -d '0' -f2 | sed 's/CatResults\/cat_//g' | sed 's/ //')

# for each gene create a gff3 file containing results and move it to HumanResults folder
for i in ${HUMANGENES[@]}
  do
  grep -i "${i/.gene.gff3/}" Homo_sapiens.GRCh38.92.gff3 > human_${i/.gene.gff3/}.gff3
  mv human_${i/.gene.gff3/}.gff3 HumanResults
done

```

#3b: Generate Human gene fasta files for BLAST - Bedtools installation required

```{r Generate Human gene fasta files for BLAST, eval=FALSE, engine='bash'}
# Remove non-gene (e.g. mRNA) strings from .gff3 results

cd $GFF_Path/HumanResults

HUMANREFINE=$(ls human_*.gff3)

for i in ${HUMANREFINE[@]}
  do
  awk -v "key=gene" '$3 == key {print($0)}' ${i} > ${i/.gff3}.gene.gff3
done


# list all cat gene coordinates files
HUMANcoords=$(ls human_*.gene.gff3)

# generate fasta file for each gene.gff3 coordinates file
for i in ${HUMANcoords[@]}
  do
  bedtools getfasta -fo "${i/.gene.gff3}".fa -fi $GFF_Path/Homo_sapiens.GRCh38.dna_sm.toplevel.fa -bed ${i}
done

```

#3c BLAST human genes to Lynx genome - BLAST installation required (from NCBI)

```{r  BLAST human genes to Lynx genome, eval=FALSE, engine='bash'}

cd $GFF_Path

# create lynx BLASTdatabase (directory and database inside) from FASTA lynx reference genome
# ALREADY DONE IN PREVIOUS STEPS

mkdir BLASTdatabase

BDB=~/home/ebazzicalupo/Data/BLASTdatabase

mv lp23_plus_mt.fa $BDB

makeblastdb -in $BDB/lp23_plus_mt.fa -parse_seqids -dbtype nucl

# BLAST procedure with new directory containing results file for each GENE (look at BLAST manual for output options)

cd $GFF_Path/HumanResults

HBLASTGENES=$(ls human_*.fa)

for GENE in ${HBLASTGENES[@]}
  do
  echo "$GENE starting BLAST"
  blastn -query $GFF_Path/HumanResults/${GENE} -db $BDB/lp23_plus_mt.fa -outfmt "6 qseqid sseqid sstart send length qcovhsp evalue" -out ${GENE/.fa/}.BLASTresults
  echo "$GENE BLAST finished"
  
done

# Move BLASTresults to new directory

mkdir BLASTresults

mv *.BLASTresults BLASTresults


```

#4a copy BLASTresults files from server to laptop

```{r  copy BLASTresults files from server to laptop, eval=FALSE, engine='bash'}

# cat greps
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/CatResults/*.gff3 ~/home/ebazzicalupo/GrepResults

# cat coordinates
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/CatResults/BLASTresults/cat_*.BLASTresults ~/home/ebazzicalupo/Results

# human greps
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/HumanResults/*.gff3 ~/home/ebazzicalupo/GrepResults

# human
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/HumanResults/BLASTresults/human_*.BLASTresults ~/home/ebazzicalupo/Results


```

#5a Create .gff3 / .bed of genes from ParentIDs

```{r  Create .gff3 of genes from ParentIDs, eval=FALSE, engine='bash'}

cd $GFF_Path/LynxResults/ParentIDs

cat * > ID.txt

mv ID.txt $GFF_Path

cd $GFF_Path

# create .gff3 file
grep -f ID.txt LYPA23C.all.fix.gff3 > Candidate.gff3

# create .bed file with: scaffold, start, end, type, info
cut -f1,3,4,5,9 Candidate.gff3 | awk 'BEGIN {FS="\t"; OFS="\t"} {print $1, $3, $4, $2, $5}' > Candidate.bed

# modify .bed to allow .vcf compatibility (scaffold name matching) - sort and insert a . at 4th character in each line
sort -k1,2 Candidate.bed | awk -vFS="" -vOFS="" '{$4=$4"."}1' > Candidate.compatible.bed


```


#5b Intersect .vcf with .bed coordinates - BCFtools

```{r  Intersect .vcf with .bed coordinates, eval=FALSE, engine='bash'}

cd $GFF_Path

mkdir VCF

# bg compression required for bcftools view
bgzip lp_species_SNPs_tagged_filtered.vcf

# indexing of .vcf file required for bcftools view
bcftools index lp_species_SNPs_tagged_filtered.vcf.gz

bcftools view -Ov -o $GFF_Path/VCF/lp_species_SNPs_filtered_intersect.vcf -R Candidate.compatible.bed lp_species_SNPs_tagged_filtered.vcf.gz

# optional uncompress file
gunzip lp_species_SNPs_tagged_filtered.vcf.gz


```


#5b Intersect .vcf with .bed/.gff3 coordinates - BEDtools

```{r  Intersect .vcf with .bed coordinates, eval=FALSE, engine='bash'}

cd $GFF_Path

mkdir VCF

bedtools intersect -header -a lp_species_SNPs_tagged_filtered.vcf.gz -b Candidate.compatible.bed > $GFF_Path/VCF/lp_bedIntersect.vcf


```

#5c Add information regarding "hit" region - BCFtools annotate

```{r  Add information regarding "hit" region - BCFtools, eval=FALSE, engine='bash'}

cd $GFF_Path

## ADD INFO REGARDING HIT REGION

# index Candidate.compatible.bed
bgzip Candidate.compatible.bed
tabix -p bed Candidate.compatible.bed.gz
mv Candidate.compatible.bed.* $GFF_Path/VCF

# create headers file as instructed by bcftools annotate

cd $GFF_Path/VCF

touch headers
nano headers

# add information from indexed bed file to vcf
bcftools annotate -a Candidate.compatible.bed.gz -h headers -c CHROM,FROM,TO,INFO/HRT,INFO/HRD lp_bedIntersect.vcf > lp_bedIntersect_annotated.vcf

```

#5d Clear duplicates from Intersect results

```{r  Clear duplicates from Intersect results, eval=FALSE, engine='bash'}

# copy header part
grep "#" lp_bedIntersect.vcf > lp_bedIntersect_unique.vcf

# copy table, remove duplicates, add it to header
grep -v "#" lp_bedIntersect.vcf | sort -u -k1,2 >> lp_bedIntersect_unique.vcf

```

