---
title: "Genes1"
author: "Enrico"
date: "24 April 2018"
output:
  html_document: default
---

#0a: Define paths.

```{r Define paths, eval=FALSE, engine='bash'}

# path to gff data in my laptop
GFF_Path=~/home/ebazzicalupo/Data

# path to gff data in server
GFF_Path=~/Data

```

#1a: Search in Annotations

```{r Search in Annotations, eval=FALSE, engine='bash'}

cd $GFF_Path

# list elements in file containing list of genes
GENES=$(cat GeneList.txt)

mkdir LynxResults

# define the end of a line as a separator for elements so that each WHOLE line in GeneList.txt will be a search entry
IFS=$'\n' 

#for each gene create a file containing results
for i in ${GENES[@]}
  do
  grep -i -E "${i}" LYPA23C.all.fix.gff3 > ${i}
  mv ${i} LynxResults
done


```

#2a: Search empty results in Cat annotations

```{r Search in Annotations, eval=FALSE, engine='bash'}

cd $GFF_Path

cd LynxResults

wc -l * | grep ' 0 ' | cut -d '0' -f2 | sed 's/ //'

# grab all empty results and print names in file in my laptop
wc -l * | grep ' 0 ' | cut -d '0' -f2 | sed 's/ //' > CatSearch.txt

# grab all empty results and print names in file in Server
wc -l * | grep ' 0 ' | cut -d '0' -f2 | sed 's/ //' > CatSearch.txt

mv CatSearch.txt $GFF_Path

cd $GFF_Path

mkdir CatResults

# list elements in file containing list of genes
CATGENES=$(cat CatSearch.txt)

cat CatSearch.txt

# define the end of a line as a separator for elements so that each WHOLE line in GeneList.txt will be a search entry
IFS=$'\n' 

# for each gene create a gff3 file containing results
for i in ${CATGENES[@]}
  do
  grep -i -E "{i}" Felis_catus.Felis_catus_8.0.92.gff3 > cat_${i}.gff3
  mv cat_${i}.gff3 CatResults
done

```

#2b: (EXAMPLE FOR 1 FILE): Generate Cat gene fasta file for BLAST - Bedtools installation required

```{r Generate Cat gene fasta file for BLAST, eval=FALSE, engine='bash'}

cd $GFF_Path

# create gene file for trial
grep "agouti" Felis_catus.Felis_catus_8.0.92.gff3 > cat_agouti.gff3

mv cat_agouti.gff3 Cat_Results

# generate fasta file (-fo) from input fasta file (-fi) using coordinates specified in other file (-bed)
bedtools getfasta -fo try.fa -fi Felis_catus.Felis_catus_8.0.dna_sm.toplevel.fa -bed cat_agouti.gff3


```

#2c: Generate Cat gene fasta files for BLAST - Bedtools installation required

```{r Generate Cat gene fasta files for BLAST, eval=FALSE, engine='bash'}

cd $GFF_Path/CatResults

# remove empty results in server

wc -l * | grep ' 0 ' | cut -d ' ' -f5 > Removelist.txt

REMlist=$(cat Removelist.txt)

for i in ${REMlist[@]}
  do
  rm ${i}
done

# list all cat gene coordinates files
CATcoords=$(ls cat_*.gff3)

# generate fasta file for each gene.gff3 coordinates file
for i in ${CATcoords[@]}
  do
  bedtools getfasta -fo "${i/.gff3}".fa -fi $GFF_Path/Felis_catus.Felis_catus_8.0.dna_sm.toplevel.fa -bed ${i}
done

```

#2d: BLAST cat genes to Lynx genome - BLAST installation required (from NCBI)

```{r  BLAST cat genes to Lynx genome, eval=FALSE, engine='bash'}

cd $GFF_Path

# create lynx BLASTdatabase (directory and database inside) from FASTA lynx reference genome

mkdir BLASTdatabase

# in my laptop
BDB=~/home/ebazzicalupo/Data/BLASTdatabase
# in server
BDB=~/Data/BLASTdatabase

mv lp23_plus_mt.fa $BDB

makeblastdb -in $BDB/lp23_plus_mt.fa -parse_seqids -dbtype nucl

# BLAST procedure with new directory containing results file for each GENE (look at BLAST manual for output options)

cd $GFF_Path/CatResults

CBLASTGENES=$(ls cat_*.fa)

for GENE in ${CBLASTGENES[@]}
  do
  echo "$GENE starting BLAST"
  blastn -query $GFF_Path/CatResults/${GENE} -db $BDB/lp23_plus_mt.fa -outfmt "6 qseqid sseqid sstart send length qcovhsp evalue" -out ${GENE/.fa/}.BLASTresults
  echo "$GENE BLAST finished"
  
done

# Move BLASTresults to new directory

mkdir BLASTresults

mv *.BLASTresults BLASTresults

```

#3a: Search cat empty results in Human annotations

```{r Search cat empty results in Human annotations, eval=FALSE, engine='bash'}

cd $GFF_Path

cd CatResults

wc -l *.gff3 | grep ' 0 ' | cut -d '0' -f2 | sed 's/cat_//' | sed 's/ //'

# grab all empty results and print names in file cutting the 'cat_' part in my laptop
wc -l *.gff3 | grep ' 0 ' | cut -d '0' -f2 | sed 's/cat_//' | sed 's/ //' > HumanSearch.txt

# grab all empty results and print names in file cutting the 'cat_' part in Server
wc -l *.gff3 | grep ' 0 ' | cut -d '0' -f2 | sed 's/cat_//g' | sed 's/ //g' > HumanSearch.txt

# if using Removelist.txt
cat Removelist.txt | sed 's/cat_//g' > HumanSearch.txt

# Before proceeding, check if 'total' line is present, if it is, remove it using nano
nano HumanSearch.txt

mv HumanSearch.txt $GFF_Path

cd $GFF_Path

mkdir HumanResults

# list elements in file containing list of genes
HUMANGENES=$(cat HumanSearch.txt)

# for each gene create a gff3 file containing results and move it to HumanResults folder
for i in ${HUMANGENES[@]}
  do
  grep -i "${i/.gff3/}" Homo_sapiens.GRCh38.92.gff3 > human_${i/.gff3/}.gff3
  mv human_${i/.gff3/}.gff3 HumanResults
done

```

#3b: Generate Human gene fasta files for BLAST - Bedtools installation required

```{r Generate Human gene fasta files for BLAST, eval=FALSE, engine='bash'}

cd $GFF_Path/HumanResults

# list all cat gene coordinates files
HUMANcoords=$(ls human_*.gff3)

# generate fasta file for each gene.gff3 coordinates file
for i in ${HUMANcoords[@]}
  do
  bedtools getfasta -fo "${i/.gff3}".fa -fi $GFF_Path/Homo_sapiens.GRCh38.dna_sm.toplevel.fa -bed ${i}
done

```

#3c BLAST human genes to Lynx genome - BLAST installation required (from NCBI)

```{r  BLAST human genes to Lynx genome, eval=FALSE, engine='bash'}

cd $GFF_Path

# create lynx BLASTdatabase (directory and database inside) from FASTA lynx reference genome
# ALREADY DONE IN PREVIOUS STEPS

mkdir BLASTdatabase

BDB=~/home/ebazzicalupo/Data/BLASTdatabase

mv lp23_plus_mt.fa $BDB

makeblastdb -in $BDB/lp23_plus_mt.fa -parse_seqids -dbtype nucl

# BLAST procedure with new directory containing results file for each GENE (look at BLAST manual for output options)

cd $GFF_Path/HumanResults

HBLASTGENES=$(ls human_*.fa)

for GENE in ${HBLASTGENES[@]}
  do
  echo "$GENE starting BLAST"
  blastn -query $GFF_Path/HumanResults/${GENE} -db $BDB/lp23_plus_mt.fa -outfmt "6 qseqid sseqid sstart send length qcovhsp evalue" -out ${GENE/.fa/}.BLASTresults
  echo "$GENE BLAST finished"
  
done

# Move BLASTresults to new directory

mkdir BLASTresults

mv *.BLASTresults BLASTresults


```

#4a copy BLASTresults files from server to laptop

```{r  BLAST human genes to Lynx genome, eval=FALSE, engine='bash'}

# cat greps
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/CatResults/*.gff3 ~/home/ebazzicalupo/GrepResults

# cat coordinates
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/CatResults/BLASTresults/cat_*.BLASTresults ~/home/ebazzicalupo/Results

# human greps
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/HumanResults/*.gff3 ~/home/ebazzicalupo/GrepResults

# human
scp ebazzicalupo@genomics-a.ebd.csic.es:~/Data/HumanResults/BLASTresults/human_*.BLASTresults ~/home/ebazzicalupo/Results


```
